### PR TITLE
Enable AffineConstraints::copy_from for mixed number (double/float)

### PR DESCRIPTION
**Pull request by peterrum**
FYI @bergbauer

### CODE REVIEW

File: include/deal.II/lac/affine_constraints.h [hunk 1]
```diff
@@ -1597,6 +1598,25 @@ class AffineConstraints : public Subscriptor
      */
     number inhomogeneity;
 
+    /**
+     * Copy constructor.
+     */
+    template <typename ConstraintLineType>
+    void
+    operator=(const ConstraintLineType &other) const
+    {
```
Comment (bangerth) [inline include/deal.II/lac/affine_constraints.h:1607]: Move it out of line.

File: include/deal.II/lac/affine_constraints.h [hunk 2]
```diff
@@ -1597,6 +1598,25 @@ class AffineConstraints : public Subscriptor
...
+      this->index = other.index;
+
+      entries.clear();
+      entries.reserve(other.entries.size());
+
+      for (const auto &i : other.entries)
+        entries.emplace_back(i.first, i.second);
```
Comment (bangerth) [inline include/deal.II/lac/affine_constraints.h:1614]: I think you can do this simpler by saying
```
  entries.clear();
  entries.insert (other.entries.begin(), other.entries.end());
```

**Review by bangerth — COMMENTED** [review]
Looks mostly ok. Might be nice to have a test. 

This can only work if you have types for the scalar type that are actually convertible into each other. It might be nice to check that with a `static_assert` so people get a reasonable error message. You could use `std::is_constructible<To,From>` in that `static_assert`.

**Comment by peterrum** [comment]
@bangerth I have addressed your comments. Ready from my side!

File: include/deal.II/lac/affine_constraints.h [hunk 3]
```diff
@@ -2467,6 +2507,49 @@ AffineConstraints<number>::add_entries_local_to_global(
...
+
+template <typename number>
+template <typename ConstraintLineType>
+inline typename AffineConstraints<number>::ConstraintLine &
+AffineConstraints<number>::ConstraintLine::
+operator=(const ConstraintLineType &other)
+{
+  this->index = other.index;
+
+  entries.insert(entries.begin(), other.entries.begin(), other.entries.end());
```
Comment (bangerth) [inline include/deal.II/lac/affine_constraints.h:2546]: I think you need to clear `entries` before the `insert` call.

File: tests/lac/constrains_copy_from_mixed.cc [hunk 4]
```diff
@@ -0,0 +1,63 @@
+// ---------------------------------------------------------------------
```
Comment (bangerth) [inline tests/lac/constrains_copy_from_mixed.cc:1]: I suspect that you meant to say "constraints" instead of "constrains" in the name of the file.

**Comment by peterrum** [comment]
@bangerth You are right! Fixed!

**Review by bangerth — APPROVED** [review]

**Pull request merged**